File: xmldiff/xmldiff/zzs.py
#!/Usr/bin/env python
# -*- coding: utf-8 -*-
# ----------------------------------------------------
# Copyright The IETF Trust 2018-9, All Rights Reserved
# ----------------------------------------------------
#  Adapted from code by
# Authors: Tim Henderson and Steve Johnson
# Email: tim.tadh@gmail.com, steve@steveasleep.com
# For licensing see the LICENSE file in the top level directory.
#
#  The original code just computed the difference and did not bother with the set of
#  edits that were needed.  We care about the edits themselves.

from __future__ import absolute_import
from six.moves import range

import collections

"""
def py_zeros(dim, pytype):
    assert len(dim) == 2
    return [[pytype for y in range(dim[1])]
            for x in range(dim[0])]
"""

try:
    from editdist import distance as strdist
except ImportError:
    def strdist(a, b):
        if a == b:
            return 0
        else:
            return 1


class EditItem(object):
    __slots__ = ['operation', 'left', 'right', 'cost']
    OP_INSERT = 1
    OP_DELETE = 2
    OP_RENAME = 3
    OP_MATCH = 4
    OP_COMBINE = 5

    def __init__(self):
        self.cost = 0
        self.operation = 0
        self.left = None
        self.right = None
        pass

    def reset(self):
        self.operation = 0
        self.cost = 0
        self.left = None
        self.right = None

    def setOperation(self, operation, left, right):
        self.operation = operation
        self.left = left
        self.right = right
        self.cost = 0
        if left is not None and isinstance(left, EditItem):
            self.cost += left.cost
        if right is not None and isinstance(right, EditItem):
            self.cost += right.cost

    def clone(self):
        cloneOut = EditItem()
        returnValue = cloneOut
        cloneOut.operation = self.operation
        cloneOut.cost = self.cost
        stack = []
        if self.left is not None:
            stack.append((1, cloneOut, self))
        if self.right is not None:
            stack.append((0, cloneOut, self))

        while len(stack) > 0:
            side, node, node1 = stack.pop()
            xx = node1.left if side else node1.right

            if xx is not None:
                if isinstance(xx, EditItem):
                    if xx.operation == 0:
                        continue
                    if xx.operation == EditItem.OP_COMBINE:
                        newNode = EditItem()
                        newNode.operation = xx.operation
                        newNode.cost = xx.cost
                        if xx.left is not None:
                            stack.append((1, newNode, xx))
                        if xx.right is not None:
                            stack.append((0, newNode, xx))
                    else:
                        newNode = xx
                else:
                    newNode = xx
                if side:
                    node.left = newNode
                else:
                    node.right = newNode

        return returnValue

    def toList(self):
        list = []
        self._toList(list)
        return list

    def _toList(self, list):
        if self.operation != EditItem.OP_COMBINE:
            list.append(self)
            return
        if self.left is not None and isinstance(self.left, EditItem):
            self.left._toList(list)
        if self.right is not None and isinstance(self.right, EditItem):
            self.right._toList(list)

    def toString(self):
        left = ""
        right = ""
        if self.left:
            left = "left= {:>3}".format(self.left.index)
        if self.right:
            right = " right= {:>3}".format(self.right.index)
        txt = "OP={0} {1:>9}{2}".format(str(self.operation), left, right)
        return txt


class AnnotatedTree(object):

    def __init__(self, root, get_children):
        self.get_children = get_children

        self.root = root
        self.nodes = list()  # a post-order enumeration of the nodes in the tree
        self.ids = list()    # a matching list of ids
        self.lmds = list()   # left most descendents
        self.keyroots = None

        # k and k' are nodes specified in the post-order enumeration.
        # keyroots = {k | there exists no k'>k such that lmd(k) == lmd(k')}
        # see paper for more on keyroots

        stack = list()
        pstack = list()
        stack.append((root, collections.deque()))
        j = 0
        while len(stack) > 0:
            n, anc = stack.pop()
            nid = j
            for c in self.get_children(n):
                a = collections.deque(anc)
                a.appendleft(nid)
                stack.append((c, a))
            pstack.append(((n, nid), anc))
            j += 1
        lmds = dict()
        keyroots = dict()
        i = 0
        while len(pstack) > 0:
            (n, nid), anc = pstack.pop()
            # print list(anc)
            self.nodes.append(n)
            self.ids.append(nid)
            # print n.label, [a.label for a in anc]
            if not self.get_children(n):
                lmd = i
                for a in anc:
                    if a not in lmds:
                        lmds[a] = i
                    else:
                        break
            else:
                try:
                    lmd = lmds[nid]
                except Exception:
                    import pdb
                    pdb.set_trace()
            self.lmds.append(lmd)
            keyroots[lmd] = i
            i += 1
        self.keyroots = sorted(keyroots.values())


def distance(A, B, get_children, insert_cost, remove_cost, update_cost):
    '''Computes the exact tree edit distance between trees A and B with a
    richer API than :py:func:`zss.simple_distance`.
    Use this function if either of these things are true:
    * The cost to insert a node is **not** equivalent to the cost of changing
      an empty node to have the new node's label
    * The cost to remove a node is **not** equivalent to the cost of changing
      it to a node with an empty label
    Otherwise, use :py:func:`zss.simple_distance`.
    :param A: The root of a tree.
    :param B: The root of a tree.
    :param get_children:
        A function ``get_children(node) == [node children]``.  Defaults to
        :py:func:`zss.Node.get_children`.
    :param insert_cost:
        A function ``insert_cost(node) == cost to insert node >= 0``.
    :param remove_cost:
        A function ``remove_cost(node) == cost to remove node >= 0``.
    :param update_cost:
        A function ``update_cost(a, b) == cost to change a into b >= 0``.
    :return: An integer distance [0, inf+)
    '''
    A, B = AnnotatedTree(A, get_children), AnnotatedTree(B, get_children)
    # treedists = zeros((len(A.nodes), len(B.nodes)), int)
    treedists = [[None for x in range(len(B.nodes))] for y in range(len(A.nodes))]
    fd = [[EditItem() for x in range(len(B.nodes)+2)] for y in range(len(A.nodes)+2)]

    A_remove = [remove_cost(A.nodes[i]) for i in range(len(A.nodes))]
    B_insert = [insert_cost(B.nodes[i]) for i in range(len(B.nodes))]

    def treedist(i, j):
        Al = A.lmds
        Bl = B.lmds
        An = A.nodes
        Bn = B.nodes

        m = i - Al[i] + 2
        n = j - Bl[j] + 2

        for x in range(1, m):
            fd1 = fd[x]
            for y in range(1, n):
                fd1[y].reset()

        ioff = Al[i] - 1
        joff = Bl[j] - 1

        for x in range(1, m):  # δ(l(i1)..i, θ) = δ(l(1i)..1-1, θ) + γ(v → λ)
            fd[x][0].setOperation(EditItem.OP_COMBINE, fd[x-1][0], A_remove[x+ioff])
        fd1 = fd[0]
        for y in range(1, n):  # δ(θ, l(j1)..j) = δ(θ, l(j1)..j-1) + γ(λ → w)
            fd1[y].setOperation(EditItem.OP_COMBINE, fd1[y-1], B_insert[y+joff])

        for x in range(1, m):  # the plus one is for the xrange impl
            fd1 = fd[x]
            fdm1 = fd[x-1]
            x_ioff = x + ioff
            t = Al[i] == Al[x_ioff]
            remove = A_remove[x_ioff]
            treedists_1 = treedists[x_ioff]
            left = An[x_ioff]
            p = Al[x_ioff]-1-ioff
            fd_p = fd[p]
            y_joff = joff
            for y in range(1, n):
                y_joff += 1
                # only need to check if x is an ancestor of i
                # and y is an ancestor of j
                if t and Bl[j] == Bl[y_joff]:
                    #                   +-
                    #                   | δ(l(i1)..i-1, l(j1)..j) + γ(v → λ)
                    # δ(F1 , F2 ) = min-+ δ(l(i1)..i , l(j1)..j-1) + γ(λ → w)
                    #                   | δ(l(i1)..i-1, l(j1)..j-1) + γ(v → w)
                    #                   +-

                    insert = B_insert[y_joff]
                    update = update_cost(left, Bn[y_joff])

                    op1Cost = fdm1[y].cost + remove.cost
                    op2Cost = fd1[y-1].cost + insert.cost
                    op3Cost = fdm1[y-1].cost + update.cost

                    if op1Cost < op2Cost:
                        if op1Cost < op3Cost:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fdm1[y], remove)
                        elif op2Cost < op3Cost:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fd1[y-1], insert)
                        else:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fdm1[y-1], update)
                    else:
                        if op2Cost < op3Cost:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fd1[y-1], insert)
                        else:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fdm1[y-1], update)

                    # fd[x][y] = min(
                    #     fd[x-1][y] + remove_cost(An[x+ioff]),
                    #     fd[x][y-1] + insert_cost(Bn[y+joff]),
                    #     fd[x-1][y-1] + update_cost(An[x+ioff], Bn[y+joff]),
                    # )
                    treedists_1[y_joff] = fd1[y].clone()
                else:
                    #                   +-
                    #                   | δ(l(i1)..i-1, l(j1)..j) + γ(v → λ)
                    # δ(F1 , F2 ) = min-+ δ(l(i1)..i , l(j1)..j-1) + γ(λ → w)
                    #                   | δ(l(i1)..l(i)-1, l(j1)..l(j)-1)
                    #                   |                     + treedist(i1,j1)
                    #                   +-
                    q = Bl[y_joff]-1-joff

                    insert = B_insert[y_joff]

                    op1Cost = fdm1[y].cost + remove.cost
                    op2Cost = fd1[y-1].cost + insert.cost
                    op3Cost = fd_p[q].cost + treedists_1[y_joff].cost

                    if op1Cost < op2Cost:
                        if op1Cost < op3Cost:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fdm1[y], remove)
                        elif op2Cost < op3Cost:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fd1[y-1], insert)
                        else:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fd_p[q], treedists_1[y_joff])
                    else:
                        if op2Cost < op3Cost:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fd1[y-1], insert)
                        else:
                            fd1[y].setOperation(EditItem.OP_COMBINE, fd_p[q], treedists_1[y_joff])

                    # fd[x][y] = min(
                    #     fd[x-1][y] + remove_cost(An[x+ioff]),
                    #     fd[x][y-1] + insert_cost(Bn[y+joff]),
                    #     fd[p][q] + treedists[x+ioff][y+joff]
                    # )

    for i in A.keyroots:
        for j in B.keyroots:
            treedist(i, j)

    return treedists[-1][-1]

File: xmldiff/xmldiff/zzs.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <assert.h>

#define _CrtCheckMemory()

int compare(const void * a, const void * b)
{
    return (*(int*)a - *(int *)b);
}

typedef struct eElement {
    int operation;
    void * left;
    void * right;
} eElement;

typedef struct eArray {
  int c;
  eElement rgEdits[];
} eArray;

typedef struct cArray {
  int c;
  void * rg[];
} cArray;


struct xxx {
    int i;
};

typedef struct intArray {
    int cItems;
    int cAlloc;
    int * rgValues;
} intArray;

typedef struct ptrArray {
    int cItems;
    int cAlloc;
    void ** rgValues;
} ptrArray;

struct snode {
    void * treeNode;
    intArray anc;
    struct snode *   pNext;
};

struct pnode {
    void * treeNode;
    int nodeId;
    intArray anc;
	int hasChildren;
    struct pnode * pNext;
};

struct aTree {
    int i;
    int       size;

    intArray  lmds;
    ptrArray  nodes;
    intArray  ids;
    intArray  keyroots;
};

void cloneIntArray(intArray * dest, intArray * src)
{
    dest->cItems = src->cItems;
    dest->cAlloc = 20;
    if (dest->cItems+20 > dest->cAlloc) {
	dest->cAlloc = dest->cItems + 5;
    }
    
    dest->rgValues = (int *) calloc(sizeof(int), dest->cAlloc);
    if (dest->rgValues == NULL) {
	//  error
    }
    memcpy(dest->rgValues, src->rgValues, src->cItems*sizeof(int));
}

void appendLeft(intArray * dest, int newValue)
{
    if (dest->cItems == dest->cAlloc) {
	int * newArray = (int *) calloc(sizeof(int), dest->cItems+10);
	memcpy(newArray, dest->rgValues, dest->cItems*sizeof(int));
	free(dest->rgValues);
	dest->rgValues = newArray;
	dest->cAlloc += 10;
    }
    memmove(&dest->rgValues[1], &dest->rgValues[0], dest->cItems*sizeof(int));
    dest->rgValues[0] = newValue;
    dest->cItems += 1;
}

void append(intArray * dest, int newValue)
{
    if (dest->cItems == dest->cAlloc) {
	int * newArray = (int *) calloc(sizeof(int), dest->cItems+10);
	memcpy(newArray, dest->rgValues, dest->cItems*sizeof(int));
	free(dest->rgValues);
	dest->rgValues = newArray;
	dest->cAlloc += 10;
    }
    dest->rgValues[dest->cItems] = newValue;
    dest->cItems += 1;
}

void appendNode(ptrArray * dest, void * newValue)
{
    if (dest->cItems == dest->cAlloc) {
	void ** newArray = (void **) calloc(sizeof(void *), dest->cItems+10);
	memcpy(newArray, dest->rgValues, dest->cItems*sizeof(void *));
	free(dest->rgValues);
	dest->rgValues = newArray;
	dest->cAlloc += 10;
    }
    dest->rgValues[dest->cItems] = newValue;
    dest->cItems += 1;
}


struct aTree * AnnotateTree(void * root, struct cArray *(*get_children)(void *))
{
	int i;
	int lmd;
	int j = 0;
	int c;
	int jj;
	struct snode * stack = NULL;
	struct pnode * pstack = NULL;
	struct aTree * self = NULL;
	int * lmds = NULL;
	int * keyroots = NULL;
	struct snode * snew = (struct snode *) calloc(sizeof(struct snode), 1);
	if (snew == NULL) {
	  return NULL;
	}
	snew->treeNode = root;
	snew->pNext = stack;
	stack = snew;

	while (stack != NULL) {
		struct snode * sthis = stack;
		int nid = j;

		struct cArray * children = get_children(sthis->treeNode);
		struct pnode * pnew;
		stack = stack->pNext;

		for (c = 0; c < children->c; c++) {
			snew = (struct snode *) calloc(sizeof(struct snode), 1);
			snew->treeNode = children->rg[c];
			cloneIntArray(&snew->anc, &sthis->anc);
			appendLeft(&snew->anc, nid);
			snew->pNext = stack;
			stack = snew;
		}
		pnew = (struct pnode *) calloc(sizeof(struct pnode), 1);
		pnew->treeNode = sthis->treeNode;
		pnew->nodeId = nid;
		pnew->anc = sthis->anc;
		pnew->hasChildren = children->c != 0;
		pnew->pNext = pstack;
		pstack = pnew;
		j += 1;

		free(sthis);
		// free(children);
	}

	self = (struct aTree *) calloc(sizeof(struct aTree), 1);
	self->size = j;

	lmds = (int *)malloc(sizeof(int) * j);
	memset(lmds, -1, sizeof(int)*j);
	keyroots = (int *)malloc(sizeof(int) * j);
	memset(keyroots, -1, sizeof(int)*j);

	i = 0;
	while (pstack != NULL) {
		struct pnode * pthis = pstack;
		pstack = pstack->pNext;

		appendNode(&self->nodes, pthis->treeNode);
		append(&self->ids, pthis->nodeId);

		if (!pthis->hasChildren) {
			int ii;
			lmd = i;
			for (ii = 0; ii < pthis->anc.cItems; ii++) {
				int a = pthis->anc.rgValues[ii];
				if (lmds[a] == -1) {
					lmds[a] = i;
				}
				else {
					break;
				}
			}
		}
		else {
			lmd = lmds[pthis->nodeId];
		}
		append(&self->lmds, lmd);
		keyroots[lmd] = i;
		i += 1;

		free(pthis->anc.rgValues);
		free(pthis);
	}

	qsort(keyroots, j, sizeof(int), compare);

	for (jj = 0; jj < j; jj++) {
		if (keyroots[jj] != -1) {
			if (self->keyroots.rgValues == NULL) {
				self->keyroots.cAlloc = j - jj;
				self->keyroots.rgValues = (int *) calloc(sizeof(int), self->keyroots.cAlloc);
			}
			self->keyroots.rgValues[self->keyroots.cItems] = keyroots[jj];
			self->keyroots.cItems += 1;
		}
	}

	free(lmds);
	free(keyroots);

	return self;
}

#define index(a, b) ((a)*n + (b))
#define tindex(x, y) ((x)*(b->size) + y)

#define OP_INSERT 1
#define OP_REMOVE 2
#define OP_UPDATE 3
#define OP_MATCH 4
#define OP_COMBINE 5
#define OP_COMBINE_MATCH 6
#define OP_COMBINE_UPDATE 7
#define OP_LIST 8

typedef struct editList {
    int  operation;
    int  cost;
    void * left;
    void * right;
	void * three;
} EditList;

void Combine(EditList * pdest, EditList * pleft, EditList * pright)
{
	pdest->operation = OP_COMBINE;
	pdest->cost = 0;
	if (pleft != NULL) {
		pdest->cost += pleft->cost;
		pdest->left = pleft;
	}
	if (pright != NULL) {
		pdest->cost += pright->cost;
		pdest->right = pright;
	}
}

void UpdateAndCombine(EditList * pdest, EditList * pleft, void * left, void * right, int cost)
{
	pdest->operation = cost ? OP_COMBINE_UPDATE : OP_COMBINE_MATCH;
	pdest->cost = cost;
	pdest->right = left;
	pdest->three = right;
	if (pleft != NULL) {
		pdest->cost += pleft->cost;
		pdest->left = pleft;
	}
}

int countList(struct eArray * psrc)
{
	int c = 0;
	int i;
	for (i = 0; i < psrc->c; i++) {
		switch (psrc->rgEdits[i].operation) {
		case OP_LIST:
			c += countList((struct eArray *) (psrc->rgEdits[i].left));
			break;

		default:
			c += 1;
		}
	}
	return c;
}

int cloneCount(EditList * p, int fAll)
{
	switch (p->operation) {
	case OP_LIST:
		if (fAll) {
			return countList((struct eArray *) ((struct cArray *) p->left));
		}
		return 1;

	case OP_INSERT:
	case OP_REMOVE:
		return 1;

	case OP_COMBINE:
		return cloneCount((EditList *) p->left, fAll) + cloneCount((EditList *) p->right, fAll);

	case OP_COMBINE_MATCH:
	case OP_COMBINE_UPDATE:
		return cloneCount((EditList *) p->left, fAll) + 1;

	default:
		return 0;
	}
}

void copyList(eArray * pdst, eArray * psrc)
{
	int i;
	for (i = 0; i < psrc->c; i++) {
		switch (psrc->rgEdits[i].operation) {
		case OP_LIST:
			copyList(pdst, (struct eArray *) psrc->rgEdits[i].left);
			break;

		default:
			pdst->rgEdits[pdst->c].operation = psrc->rgEdits[i].operation;
			pdst->rgEdits[pdst->c].left = psrc->rgEdits[i].left;
			pdst->rgEdits[pdst->c].right = psrc->rgEdits[i].right;
			pdst->c += 1;
			break;
		}
	}
}

void fillArray(eArray * pa, EditList * p, int fAll)
{
	switch (p->operation) {
	case OP_LIST:
		if (fAll) {
			struct eArray * pca = (struct eArray *) p->left;
			copyList(pa, pca);
			return;
		}
		pa->rgEdits[pa->c].operation = OP_LIST;
		pa->rgEdits[pa->c].left = p->left;
		pa->c += 1;
		return;

	case OP_INSERT:
	case OP_REMOVE:
		pa->rgEdits[pa->c].operation = p->operation;
		pa->rgEdits[pa->c].left = p->left;
		pa->rgEdits[pa->c].right = p->right;
		pa->c += 1;
		return;

	case OP_COMBINE:
		fillArray(pa, (EditList *) p->left, fAll);
		fillArray(pa, (EditList *) p->right, fAll);
		return;

	case OP_COMBINE_MATCH:
	case OP_COMBINE_UPDATE:
		fillArray(pa, (EditList *) p->left, fAll);
		pa->rgEdits[pa->c].operation = p->operation == OP_COMBINE_MATCH ? OP_MATCH : OP_UPDATE;
		pa->rgEdits[pa->c].left = p->right;
		pa->rgEdits[pa->c].right = p->three;
		pa->c += 1;
		return;

	default:
		return;
	}
}

EditList * cloneEdits(EditList * p2clone, int fAll)
{
	int c = cloneCount(p2clone, fAll);
	EditList * pret = NULL;
	
	eArray * pEA = (eArray *)calloc(sizeof(eArray) + sizeof(eElement) * c, 1);
	pEA->c = 0;
	fillArray(pEA, p2clone, fAll);

	pret = (EditList *)calloc(sizeof(EditList), 1);
	pret->operation = OP_LIST;
	pret->left = pEA;
	pret->cost = p2clone->cost;

	return pret;
}

struct eArray * Distance(void * leftTree, void * rightTree, struct cArray *(*get_children)(void *),
	int(*insert_cost)(void *), int(*remove_cost)(void *),
	int(*update_cost)(void *, void *))
{
	struct aTree * a = AnnotateTree(leftTree, get_children);
	struct aTree * b = AnnotateTree(rightTree, get_children);

	EditList ** treedists = (EditList **)calloc(sizeof(EditList *), (a->size)*(b->size));
	EditList * a_remove = (EditList *)calloc(sizeof(EditList), (a->size));
	EditList * b_insert = (EditList *)calloc(sizeof(EditList), (b->size));

	EditList * fd = (EditList *)malloc(sizeof(EditList) * (a->size+1) * (b->size+1));
	EditList * pret = NULL;
	eArray * pret2;
	
	int i;
	int i_1;

	for (i = 0; i < a->size; i++) {
		a_remove[i].operation = OP_REMOVE;
		a_remove[i].cost = remove_cost(a->nodes.rgValues[i]);
		a_remove[i].left = a->nodes.rgValues[i];
	}

	for (i = 0; i < b->size; i++) {
		b_insert[i].operation = OP_INSERT;
		b_insert[i].cost = insert_cost(b->nodes.rgValues[i]);
		b_insert[i].right = b->nodes.rgValues[i];
	}

	for (i_1 = 0; i_1 < a->keyroots.cItems; i_1++) {
		int i = a->keyroots.rgValues[i_1];
		// _CrtCheckMemory();
		int j_1;
		for (j_1 = 0; j_1 < b->keyroots.cItems; j_1++) {
			int j = b->keyroots.rgValues[j_1];
			int m = i - a->lmds.rgValues[i] + 2;
			int n = j - b->lmds.rgValues[j] + 2;
			int ioff = a->lmds.rgValues[i] - 1;
			int joff = b->lmds.rgValues[j] - 1;
			int x;
			int y;
			
			_CrtCheckMemory();

			memset(fd, 0, sizeof(EditList)*m*n);
			assert(treedists != NULL);
			_CrtCheckMemory();


			_CrtCheckMemory();
			
			for (x = 1; x < m; x++) {
				Combine(&fd[index(x, 0)], &fd[index(x - 1, 0)], &a_remove[x + ioff]);
			}
			_CrtCheckMemory();

			for (y = 1; y < n; y++) {
				Combine(&fd[index(0, y)], &fd[index(0, y - 1)], &b_insert[y + joff]);
			}

			for (x = 1; x < m; x++) {
				for (y = 1; y < n; y++) {
				  // _CrtCheckMemory();
					int x_ioff = x + ioff;
					int y_joff = y + joff;

					if (a->lmds.rgValues[i] == a->lmds.rgValues[x_ioff] && b->lmds.rgValues[j] == b->lmds.rgValues[y_joff]) {
						//                    +-
						//                    | δ(l(i1)..i-1, l(j1)..j) + γ(v → λ)
						//  δ(F1 , F2 ) = min-+ δ(l(i1)..i , l(j1)..j-1) + γ(λ → w)
						//                    | δ(l(i1)..i-1, l(j1)..j-1) + γ(v → w)
						//                    +-

						int cost2Update = update_cost(a->nodes.rgValues[x_ioff], b->nodes.rgValues[y_joff]);

						int op1Cost = fd[index(x - 1, y)].cost + a_remove[x_ioff].cost;
						int op2Cost = fd[index(x, y - 1)].cost + b_insert[y_joff].cost;
						int op3Cost = fd[index(x - 1, y - 1)].cost + cost2Update;

						if (op1Cost < op2Cost) {
							if (op1Cost < op3Cost) {
								Combine(&fd[index(x, y)], &fd[index(x - 1, y)], &a_remove[x_ioff]);
							}
							else if (op2Cost < op3Cost) {
								Combine(&fd[index(x, y)], &fd[index(x, y - 1)], &b_insert[y_joff]);
							}
							else {
								UpdateAndCombine(&fd[index(x, y)], &fd[index(x - 1, y - 1)], a->nodes.rgValues[x_ioff], b->nodes.rgValues[y_joff], cost2Update);
							}
						}
						else {
							if (op2Cost < op3Cost) {
								Combine(&fd[index(x, y)], &fd[index(x, y - 1)], &b_insert[y_joff]);
							}
							else {
								UpdateAndCombine(&fd[index(x, y)], &fd[index(x - 1, y - 1)], a->nodes.rgValues[x_ioff], b->nodes.rgValues[y_joff], cost2Update);
							}
						}
						_CrtCheckMemory();

						treedists[tindex(x_ioff, y_joff)] = cloneEdits(&fd[index(x, y)], 0);
						_CrtCheckMemory();
					}
					else {
						// #                   +-
						// #                   | δ(l(i1)..i-1, l(j1)..j) + γ(v → λ)
						// # δ(F1, F2) = min - +δ(l(i1)..i, l(j1)..j - 1) + γ(λ → w)
						// #                   | δ(l(i1)..l(i)-1, l(j1)..l(j)-1)
						// #                   |                     + treedist(i1,j1)
						// #                   +-

						int p = a->lmds.rgValues[x_ioff] - 1 - ioff;
						int q = b->lmds.rgValues[y_joff] - 1 - joff;

						int op1Cost = fd[index(x - 1, y)].cost + a_remove[x_ioff].cost;
						int op2Cost = fd[index(x, y - 1)].cost + b_insert[y_joff].cost;
					  int op3Cost = fd[index(p, q)].cost + ((treedists[tindex(x_ioff, y_joff)] != NULL) ? treedists[tindex(x_ioff, y_joff)]->cost : 0);

						if (op1Cost < op2Cost) {
							if (op1Cost < op3Cost) {
								Combine(&fd[index(x, y)], &fd[index(x - 1, y)], &a_remove[x_ioff]);
							}
							else if (op2Cost < op3Cost) {
								Combine(&fd[index(x, y)], &fd[index(x, y - 1)], &b_insert[y_joff]);
							}
							else {
								Combine(&fd[index(x, y)], &fd[index(p, q)], treedists[tindex(x_ioff, y_joff)]);
							}
						}
						else {
							if (op2Cost < op3Cost) {
								Combine(&fd[index(x, y)], &fd[index(x, y - 1)], &b_insert[y_joff]);
							}
							else {
								Combine(&fd[index(x, y)], &fd[index(p, q)], treedists[tindex(x_ioff, y_joff)]);
							}
						}
						_CrtCheckMemory();
					}
				}
			}
		}
	}
	_CrtCheckMemory();

	free(fd);

	pret = cloneEdits(treedists[tindex(a->size - 1, b->size - 1)], 1);

	for (i=0; i<a->size*b->size; i++) {
	  if (treedists[i]->operation == OP_LIST) {
	    free(treedists[i]->left);
	    free(treedists[i]);
	  }
	}
	free(treedists);
	free(a_remove);
	free(b_insert);

	pret2 = pret->left;
	free(pret);
	// M00TODO finish freeing everything
	return pret2;
}
